rfc/rbac-ng.md

```diff
@@ -0,0 +1,154 @@
+- Add Open Policy Agent as a storage and permission evaluator backend for Fiat
+- Add support for role policies to Kayenta metric store accounts
+- Update Front50 to push application permission changes to Fiat
+
+Non Goals:
+- Provide APIs for managing security policies directly via Rego or other potential policy backends
+
+## Motivation and Rationale
+
+Kayenta APIs are all provided fairly openly assuming an admin has configured it with usable account credentials.
+Account credentials can be configured for metric stores and object stores, the latter of which can also be used for configuration stores.
+Object stores are used for storing and retrieving metric query results which are each assigned random UUIDs upon creation.
+Configurations are encouraged to be shared and are viewable by all applications by default; as such, these are the only object stores that list objects from the store.
+It is metric stores where lack of access controls causes issues in multi-tenant environments as sources for metrics are typically guarded by team-specific credentials.
+All these accounts are defined as Spring beans through configuration files provided by the Spinnaker admin.
+
+Supporting self-service configuration of metric store accounts requires some form of access controls as metric stores may contain sensitive data.
+A self-service API should learn from the equivalent feature in Clouddriver how to integrate with user secrets and manage account definitions to provide a similar API for Kayenta.
+Users will then be able to configure Kayenta accounts at runtime rather than requiring a Spinnaker admin to update Kayenta's configuration each time.
+Multi-tenant usage of Kayenta will become feasible.
+
+## Timeline
+
+The initial Kayenta and dependent work is being developed during 2023Q1.
+Further integration with other Spinnaker microservices, role policies, and permission change tracking would be targeted for afterwards in 2023.
+
+## Design
+
+This consists of a few related security elements: authentication, authorization, and credentials management.
+Authentication is handled by Gate typically, so this will reuse the existing `X-Spinnaker-User` header for denoting the principal of the request.
+User details are tracked by Fiat, so this data will need to be mirrored into OPA as a new backend.
+Authorization is the more complex part as permission data is stored by the various microservices that own the resources.
+These checks are typically handled through the `PermissionEvaluator` interface from Spring Security which takes three arguments: the authenticated user, the resource, and the permission.
+Permission evaluation will be given the context of what roles the requesting user has, what permissions apply to applications, and what permissions apply to accounts.
+As applications are checked by multiple services, these permissions will be pushed by Front50 into Fiat so that any service can use application permissions.
```

Comment:
Does `will be pushed by Front50 into Fiat` mean a migration of existing perms from front50 data stores to fiat?

Reply:
Front50 will keep ownership of the data; it will mirror that data into Fiat for its own policy evaluation usage (typically stored ephemerally).

Reply:
For permission data, instead of Fiat periodically polling all the other services for permissions, those services will be updated to push permission data updates to Fiat through a new data collection endpoint. Fiat can make an initial call to the various services to get a current snapshot of permissions when its own cache is empty (or to reset the cache) and subsequently listen for permissions updates over time. This will allow us to keep a fairly up to date but eventually consistent cache of Front50 resource permissions, Clouddriver resource permissions, and any other services (Igor and Kayenta come to mind; maybe even Echo) with a couple advantages over the existing approach:

1. By listening to what is effectively an event stream, we keep a more up to date view of permissions than periodic polling.
2. By reducing polling to cold starts, we significantly reduce the network traffic Fiat generates.

A corresponding update to `FiatPermissionEvaluator` will allow for asking Fiat to do policy evaluation.